### DISCUSSION TITLE
How to refresh datagrid ui in avalonia

### DISCUSSION THREAD

**Discussion by keynes1981** [AvaloniaUI/Avalonia#15717] [Q&A] [answered]
Hello, I'm a beginner of Avalonia. 
 I used a button command in my ViewModel to update the ObservableCollection datasource of Datagrid, but it doesn't refresh the datagrid, What's the correct steps? Here is my code:

```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
		xmlns:vm="using:AvaloniaTest.ViewModels"
		x:DataType="vm:MainViewModel"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaTest.MainWindow"
        Title="AvaloniaTest">
	<StackPanel>
		<Button Margin="20 20 0 0" Content="ChangeData"  Command="{Binding ChangeDataCommand}"/>
		<DataGrid Margin="20 10 0 0" ItemsSource="{Binding Tasks,Mode=TwoWay}"
			  IsReadOnly="True"
			  CanUserReorderColumns="True"
			  CanUserResizeColumns="True"
			  CanUserSortColumns="False"
			  GridLinesVisibility="All"
			  BorderThickness="1" BorderBrush="Gray"  Height="200">
			<DataGrid.Columns>
				<DataGridCheckBoxColumn Header="select" Binding="{Binding IsDone}" />
				<DataGridTextColumn Header="Title" Binding="{Binding Title}" />
			</DataGrid.Columns>
		</DataGrid>
	</StackPanel>
</Window>
```

`namespace AvaloniaTest.Models
{
    public class MyTask
    {
        public string Title { get; set; }
        public bool IsDone { get; set; }

        public MyTask(string title, bool isDone)
        {
            Title = title;
            IsDone = isDone;
        }
    }
}`

`namespace AvaloniaTest.ViewModels
{
    public class MainViewModel : ReactiveObject
    {

        public ICommand? ChangeDataCommand { get; }
        private ObservableCollection<MyTask> _tasks = new();
        public ObservableCollection<MyTask> Tasks
        {
            get => _tasks;
            set => this.RaiseAndSetIfChanged(ref _tasks, value);
        }

        public MainViewModel()
        {
            ChangeDataCommand = ReactiveCommand.Create(() =>
            {
                var tasks1 = new List<MyTask>
            {
                new ("Tas1k1",false),
                new ("Task22",true),
                new ("Task3",false)
            };
                _tasks = new ObservableCollection<MyTask>(tasks1);

            });

            var tasks = new List<MyTask>
            {
                new ("Task1",true),
                new ("Task2",true),
                new ("Task3",true)
            };
            _tasks = new ObservableCollection<MyTask>(tasks);
        }
    }
}`

[AvaloniaTest.zip](https://github.com/AvaloniaUI/Avalonia/files/15306214/AvaloniaTest.zip)

**Comment by Pratham95VP** [comment, marked as answer]
` _tasks = new ObservableCollection<MyTask>(tasks1);`  replace this line with `Tasks = new ObservableCollection<MyTask>(tasks1);`

**Reply by Pratham95VP** [reply]
You are updating the _tasks field instead of the Tasks property in your ChangeDataCommand. The Tasks property is what's bound to the ItemsSource of your DataGrid, so you need to update that property in order for the changes to be reflected.

**Reply by thevortexcloud** [reply]
This is also not quiet right either. It will probably work but it's not very efficient. Instead of creating a new collection, just add the items to the existing collection. The change event only fires when the contents change.

**Reply by timunie** [reply]
best option is to use `Tasks.Add` or `Tasks.Remove` to update the collection. Only replace it if really needed.
